edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/brokerConnection/MqttBrokerConnector.cs

```diff
@@ -0,0 +1,423 @@
+            this.publications = Option.None<Channel<MqttPublishInfo>>();
+        }
+
+        // these are statics, so they don't use the state to acquire 'client' - making easier to handle parallel
+        // Reconnect/Disconnect cases
+        static async Task<bool> TryConnectAsync(MqttClient client, IReadOnlyCollection<IMessageConsumer> subscribers, string id)
+        {
+            try
+            {
+                var result = client.Connect(id, id, string.Empty);
+
+                if (result != MqttMsgConnack.CONN_ACCEPTED)
+                {
+                    Events.MqttServerDenied();
+                    throw new Exception("Mqtt server rejected bridge connection");
+                }
+
+                await SubscribeAsync(client, subscribers);
+
+                return true;
+            }
+            catch (Exception e)
+            {
+                Events.FailedToConnect(e);
+
+                // if the connection was successful at some level,
+                // try to clean up with Disconnect()
+                if (client.IsConnected)
+                {
+                    try
+                    {
+                        client.Disconnect();
+                    }
+                    catch
+                    {
+                        // swallow intentionally
+                    }
+                }
+
+                return false;
+            }
+        }
+
+        static async Task SubscribeAsync(MqttClient client, IReadOnlyCollection<IMessageConsumer> subscribers)
+        {
+            var expectedAckCount = subscribers.Count;
+
+            if (expectedAckCount > 0)
+            {
+                using (var acksArrived = new SemaphoreSlim(0, 1))
+                {
+                    var allQosGranted = default(bool);
+
+                    client.MqttMsgSubscribed += ConfirmSubscribe;
+
+                    foreach (var subscriber in subscribers)
+                    {
+                        client.Subscribe(
+                            subscriber.Subscriptions.ToArray(),
+                            Enumerable.Range(1, subscriber.Subscriptions.Count).Select(i => (byte)1).ToArray());
+                    }
+
+                    await acksArrived.WaitAsync();
+
+                    client.MqttMsgSubscribed -= ConfirmSubscribe;
+
+                    if (!Volatile.Read(ref allQosGranted))
+                    {
+                        Events.QosMismatch();
+                        throw new Exception("MQTT server did not grant QoS-1 for every requested subscription");
+                    }
+
+                    void ConfirmSubscribe(object sender, MqttMsgSubscribedEventArgs e)
+                    {
+                        if (Interlocked.Decrement(ref expectedAckCount) == 0)
+                        {
+                            Volatile.Write(ref allQosGranted, e.GrantedQoSLevels.All(qos => qos == 1));
+
+                            acksArrived.Release();
```

Comment:
What if we never receive an ack? Will this stuck? We can try to use CancellationTokenSource with timeout instead of SemaphoreSlim.